in_toto/artifact_rules.py

```diff
@@ -0,0 +1,171 @@
+
+  rule_type = rule[0].lower()
+  pattern = rule[1]
+
+  # Type is one of "CREATE", "MODIFY", "DELETE", "ALLOW", "DISALLOW"
+  if rule_type in GENERIC_RULES:
+    if rule_len != 2:
+      raise securesystemslib.exceptions.FormatError("Wrong rule format,"
+        " generic rules must have one of the formats:\n\t"
+        "CREATE <pattern>\n\t"
+        "MODIFY <pattern>\n\t"
+        "DELETE <pattern>\n\t"
+        "ALLOW <pattern>\n\t"
+        "DISALLOW <pattern>\n"
+        "Got:\n\t{}".format(rule))
+    else:
+      return {
+        "type": rule_type,
+        "pattern": pattern,
+      }
+
+  # Type is "MATCH"
+  elif rule_type in COMPLEX_RULES:
+
+    # ... IN <source-path-prefix> WITH (MATERIALS|PRODUCTS)
+    # IN <destination-path-prefix> FROM <step>
+    if (rule_len == 10 and rule[2].lower() == "in" and
+        rule[4].lower() == "with" and rule[6].lower() == "in" and
+        rule[8].lower() == "from"):
+      source_prefix = rule[3]
+      dest_type = rule[5]
+      dest_prefix = rule[7]
+      dest_name = rule[9]
+
+    # ... IN <source-path-prefix> WITH (MATERIALS|PRODUCTS) FROM <step>
+    elif (rule_len == 8 and rule[2].lower() == "in" and
+        rule[4].lower() == "with" and rule[6].lower() == "from"):
+      source_prefix = rule[3]
+      dest_type = rule[5]
+      dest_prefix = ""
+      dest_name = rule[7]
+
+    # ... WITH (MATERIALS|PRODUCTS) IN <destination-path-prefix> FROM <step>
+    elif (rule_len == 8 and rule[2].lower() == "with" and
+        rule[4].lower() == "in" and rule[6].lower() == "from"):
+      source_prefix = ""
+      dest_type = rule[3]
+      dest_prefix = rule[5]
+      dest_name = rule[7]
+
+    # ... WITH (MATERIALS|PRODUCTS) FROM <step>
+    elif (rule_len == 6 and rule[2].lower() == "with" and
+        rule[4].lower() == "from"):
+      source_prefix = ""
+      dest_type = rule[3]
+      dest_prefix = ""
+      dest_name = rule[5]
+
+    else:
+      raise securesystemslib.exceptions.FormatError("Wrong rule format,"
+          " match rules must have the format:\n\t"
+          " MATCH <pattern> [IN <source-path-prefix>] WITH"
+          " (MATERIALS|PRODUCTS) [IN <destination-path-prefix>] FROM <step>.\n"
+          "Got: \n\t{}".format(rule))
+
+    if dest_type.lower() != "materials" and dest_type.lower() != "products":
+      raise securesystemslib.exceptions.FormatError("Wrong rule format,"
+          " match rules must have either MATERIALS or PRODUCTS (case"
+          " insensitive) as destination.\n"
+          "Got: \n\t{}".format(rule))
+
+    return {
+      "type": rule_type,
+      "pattern": pattern,
+      "source_prefix": source_prefix,
+      "dest_prefix": dest_prefix,
+      "dest_type" : dest_type.lower(),
+      "dest_name": dest_name
+    }
+
+  #else:
+  # can't reach else
+  # if it is neither in GENERIC_RULES nor in MATCH_RULES it would
+  # have already raised an exception above when checking if it is in ALL_RULES
```

Comment:
I guess this else is left for the sake of documentation, but do we really want it?